crates/core/tedge_agent/src/file_transfer_server/error.rs

```diff
@@ -24,8 +23,85 @@ pub enum FileTransferError {
     BindingAddressInUse { address: std::net::SocketAddr },
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum FileTransferRequestError {
+    #[error(transparent)]
+    FromIo(#[from] std::io::Error),
+
+    #[error("Request to delete {path:?} failed: {err}")]
+    DeleteIoError {
+        #[source]
+        err: std::io::Error,
+        path: RequestPath,
+    },
+
+    #[error("Request to upload to {path:?} failed: {err:?}")]
+    Upload {
+        #[source]
+        err: anyhow::Error,
+        path: RequestPath,
+    },
+
+    #[error("Invalid file path: {path:?}")]
+    InvalidPath { path: RequestPath },
+
+    #[error("File not found: {0:?}")]
+    FileNotFound(RequestPath),
+
+    #[error("Path rejection: {0:?}")]
+    PathRejection(#[from] PathRejection),
+}
+
 impl From<FileTransferError> for RuntimeError {
     fn from(error: FileTransferError) -> Self {
         RuntimeError::ActorError(Box::new(error))
     }
 }
+
+impl IntoResponse for FileTransferError {
+    fn into_response(self) -> axum::response::Response {
+        use FileTransferError::*;
+        let status_code = match self {
+            FromIo(_)
+            | FromHyperError(_)
+            | FromAddressParseError(_)
+            | FromUtf8Error(_)
+            | BindingAddressInUse { .. } => {
+                tracing::error!("{self}");
+                StatusCode::INTERNAL_SERVER_ERROR
+            }
+        };
+        status_code.into_response()
+    }
+}
+
+impl IntoResponse for FileTransferRequestError {
+    fn into_response(self) -> axum::response::Response {
+        use FileTransferRequestError::*;
+        match &self {
+            FromIo(_) | PathRejection(_) => {
+                tracing::error!("{self}");
+                (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    "Internal error".to_owned(),
+                )
+            }
+            DeleteIoError { path, .. } => {
+                tracing::error!("{self}");
+                (
+                    StatusCode::FORBIDDEN,
```

Comment:
Do these really want to be `403 Forbidden` (it's what the code did previously)? It feels like an encapsulation failure to recognise for certain paths that we can't delete/upload to them.

Reply:
IMO "could not delete path due to IO error" could happen due e.g. path not existing, in which case we need to return 404, or due to not having permissions, in which case we should return 500, so if I'm not wrong, then even using a single return code for all IO errors is kinda bad. As for `403 Forbidden`, I also feel that it should be used when the client that made the request doesn't have permissions to do something, which doesn't seem to be the case here.

Reply:
Yeah, that was roughly what I was thinking. We currently handle the not found case by returning `202 Accepted`, like the successful deletion case (which I assume is to make `DELETE` requests idempotent).  Returning `404 Not Found` may also make sense, or if the file previously existed, `410 Gone`, although a success response tells the user the file definitely no longer exists, and the request definitely succeeded.

I think if we've hit the case handled by this code, it's some error accessing the file system, which in my mind should definitely be a `500 Internal Server Error`.

Reply:
I don't understand what application logic is relevant here. With the current API at least, we have no control over permissions. If there's a permissions error on the file system, that implies some other process is interfering with /var/tedge/file-transfer, which is not an application-level error, it's an error from the environment. The question here mostly boils down to whether we want to assume it's some file that should have never been there (in which case encapsulating the error entirely and responding with 404 Not Found is appropriate, or whether we assume the file was uploaded correctly, but something as changed the internal permissions so we can't access it, in which case a 500 Internal Server Error is appropriate.

Thinking more broadly, a useful litmus test to apply here would be "if one user receives a 403 for a resource, there must exist some other user that is able to access the resource in question". If not, and the 403 response is global, then I don't believe this can be considered an authorization issue.

Reply:
@didier-wenzek @reubenmiller do you have opinions on this point?

Reply:
Deleting a resource that doesn't exist is definitely an error. Returning 202 in such a case makes the request idempotent.

Trying to delete a non-empty directory, should be considered as a user error as we don't want to support recursive deletes. So 400 makes sense. I would do the same for empty-directories.

If the error is due to missing access, I would say this is a 500 as the file-transfer service should own all the files under file transfer.

Reply:
> 403 doesn't look bad, as [the spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403) mentions the following: `The access is tied to the application logic, such as insufficient rights to a resource.`

From RFC 9110:

> [The 403 (Forbidden) status code indicates that the server understood the request but refuses to fulfill it.](https://www.rfc-editor.org/rfc/rfc9110#section-15.5.4-1)

If we get an IO error due to missing file permissions, I would say that the server tried to execute the request, but failed, due to a misconfiguration (e.g. not having write access to `file-transfer` directory), in which case 500 would be fitting.

Reply:
> Trying to delete a non-empty directory, should be considered as a user error as we don't want to support recursive deletes. So 400 makes sense. I would do the same for empty-directories.

I think if it is a directory, 404 makes more sense as it the error is in the URL and the issue is that the resource in question cannot be interacted with (and essentially doesn't exist), not that the request is inherently invalid. I'll make sure the error message clarifies that the file we attempted to delete is a directory.